### ISSUE TITLE
Brightness function of the ColoEq is (again) strongly dependent on the saturation

### ISSUE THREAD

**Issue by s7habo** [darktable-org/darktable#16342]
### Describe the bug

I don't know if it's because of the recent changes in the module, or I've found a specific example, but the brightness again seems to be very much dependent on the saturation. 

In this example I want to darken the green:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/a2503a6c-076f-4b01-b939-afb7271b205b)

If I do this with the guided filter turned on, I get this result:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/182d925a-c072-4730-8f10-a5e9b55a7dbb)

Note that a very large area was simply not taken into account. 
It is not much better if you switch off the guided filter:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/d7b9fbda-b30b-4d8c-bc37-0d928425bafa)

 The mask also clearly shows that saturation (and brightness?) oriented selection:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/1923b81a-7a61-4758-bd8d-752a62aed606)

The result looks _much better_ if you correct the saturation beforehand with the color balance RGB module. Note that I have adjusted the saturation here so that I have saturated the highlights the most, the midtones much less and I even had to _reduce_ the saturation for shadows. Then you get the result as I would expect:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/8a274982-64e7-4be9-b7f4-aee2587662bb)

In this way, the result comes pretty close to the result if you mask the green with the exposure module and darken it:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/be1a8d11-8be6-4e40-be29-3ea9a307ec68)

EDIT: 
Just as a thought, could it be that the brightness function depends on the saturation (purity) _and_ on the _original brightness_ of the area? This means that not only the _saturated areas_, but also the _darker areas are affected much more_ than the brighter ones?

### Steps to reproduce

1. Lighten or darken a color range with different saturation/brightness in color equalizer. Depending on the saturation/brightness, the range is lightened/darkened differently, depending on the original brightness and/or saturation. The darker and more saturated areas are affected more than lighter and less saturated areas.

### Expected behavior

The affected area should be lightened/darkened evenly regardless of the original brightness/saturation of the area. 

### Where did you obtain darktable from?

self compiled

### darktable version

4.7.0+611~g76e4f8a6c

### What OS are you using?

Linux

### What is the version of your OS?

Ubuntu Studio 22.04

### Describe your system?

KDE Plasma Version: 5.24.7
KDE Frameworks Version: 5.92.0
Qt Version: 5.15.3
Kernel Version: 5.19.0-1028-lowlatency (64-bit)
Graphics Platform: X11
Processors: 8 × Intel® Core™ i7-7700K CPU @ 4.20GHz
Memory: 31.3 GiB of RAM

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

Graphics Processor: NVIDIA GeForce GTX 1070/PCIe/SSE2

**Comment by jenshannoschwalm** [comment]
I guess you are aware of the analyse radius button? You will have false colors for parts that **can** be changed in bright read and that **wont** be changed in dark blue ...

The color maths including the guided filter have not changed for "some time" - still looking for examples where it fails as in this image...

**Comment by s7habo** [comment]
> I guess you are aware of the analyse radius button? You will have false colors for parts that can be changed in bright read and that wont be changed in dark blue ...

Yes, I am aware of that. For the example above, it looks like this:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/4d6aa6ac-5489-4191-82a8-1c311a0b0a6a)

Moving the slider does not cause any change and has no effect on the result, regardless of whether the guided filter is switched on or off.

**Comment by jenshannoschwalm** [comment]
> Moving the slider does not cause any change and has no effect on the result, regardless of whether the guided filter is switched on or off.

Right. Right now it just displays the weighing mask. We still have not found a "perfect" solution as we clearly see here, often in the green tones ...

**Comment by s7habo** [comment]
It works well for saturation. However, there is a big difference between switching the guided filter on and off.

GF off:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/711d2e16-c2cf-4b3f-9d19-8725cbf58910)

GF on:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/35953dfc-c022-4fd2-898b-4ab75a9999ec)

**Comment by s7habo** [comment]
By the way, the raw file above is from 
https://www.signatureedits.com/free-raw-photos/

 Simply scroll down until it appears or enter "DSC02363" in the search field and then it will be selected and can be downloaded

![grafik](https://github.com/darktable-org/darktable/assets/38165484/2b368752-e510-44e4-8a56-268a60f1cd81)

**Comment by SoupyGit** [comment]
Yes while recent changes improved the result a lot it is still not perfect, with light colours and in particular green. 

Saturation:
![saturation](https://github.com/darktable-org/darktable/assets/69561443/fad662ec-8c53-4495-82a8-b664a0c1901b)

Brightness:
![brightness](https://github.com/darktable-org/darktable/assets/69561443/66dfe81f-7fb2-419a-8659-1ca3be5f6cc8)

**Comment by jenshannoschwalm** [comment]
@s7habo i certainly have the described problem "on radar".

Atm we have
1. dark areas being favoured over bright parts regarding "guided correction". (The earlier issue was "dark areas not corrected :-) The reason for that is known - not sure yet how to handle that "the best way"
2. greens being more problematic than other hues. That issue is still not understood.

I would love to collect all known issues with this module before starting a new iteration of work on module maths. Are you - or is anybody else - aware of other issues?

**Comment by s7habo** [comment]
> Are you - or is anybody else - aware of other issues?

Yes, there is one that immediately comes to mind.

With smooth transitions, guided filter with _brightness and saturation_ function produces banding:

With GF:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/c4bd7eaa-8a18-4254-aa80-11bcf12d0c89)

Without GF:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/e8a6f794-2408-4be1-96e7-34f449960207)

Hue function works correctly:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/984714e3-0617-4de5-9b48-619f0e920425)


I am currently testing the module in more detail. If I notice anything, I'll let you know straight away.

**Comment by jenshannoschwalm** [comment]
Hi, #16363 will fix most issues as far as I can oversee it - tested on many images you provided and others. Let's keep track of problems here after that gets merged. It should work much better for all colorful areas, dark and bright.

**Comment by jenshannoschwalm** [comment]
Beside whatever issues you'll report, I am working on suppression of halos and increased chroma noise.

**Comment by s7habo** [comment]
Weighing function is now super! 

![grafik](https://github.com/darktable-org/darktable/assets/38165484/8ac3d7fb-ff8d-471b-8012-81e74f26d28e)


The problems with bending are also gone. :clap: :+1:


The only thing that is still weak is the problem with green:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/5b717ea0-00f7-411e-86bf-f5c58819a223)

But what works very well now is the analysis radius and the guided filter.  :rocket:

**Comment by s7habo** [comment]
I have now tried a lot of photos, and now I'm not sure if it's because of green.

When I compare the masks between the saturation and brightness functions, two things stand out:

1. brightness is much weaker than saturation.
2. saturation seemed to work more **linear (flatter)** and brightness **logarithmic (more contrasty)**

Here are a few examples for comparison.

![grafik](https://github.com/darktable-org/darktable/assets/38165484/6809f059-7d76-45d9-87d6-ccb80b171b4a)

Saturation mask:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/a5c8c053-db4d-43ac-b65b-e0b766637818)


Brightness mask:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/c9b37189-779b-472b-8694-6caf947e5b21)

Org:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/173dec55-ef97-423b-a426-4a2a6054368e)

Saturation mask:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/76a75619-260b-463a-9a04-5ab1c365ec74)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/1f638a8e-680b-4f3a-b537-3c959144d7a5)


Org:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/bdbe203f-2676-41bd-81c5-19195d09dc42)

Saturation mask:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/befbab00-4c9b-4467-973c-8c575b7ff89f)

Brightness mask:
![grafik](https://github.com/darktable-org/darktable/assets/38165484/e5ed98b3-21c4-4b29-8b2a-f8c93a5e15fc)

**Comment by s7habo** [comment]
In any case, thank you very much for your efforts!

Once this little rest of the problems are cleaned up, this will be a new milestone in image processing with darktable.

**Comment by SoupyGit** [comment]
Saturation is now perfect!

![saturation](https://github.com/darktable-org/darktable/assets/69561443/5db95267-0bf6-4a0c-9143-8079c332275f)

And brightness has improved:

![brightness](https://github.com/darktable-org/darktable/assets/69561443/51a2b0d4-5f4d-44a8-978c-a76c534649ac)

> now I'm not sure if it's because of green.

Yes it seems to me now that lightness is the main determining factor. The colours that go to black quickest are those that appear darkest to begin with (red > blue). Of course I am of no help in determining the cause. Great work again @jenshannoschwalm!

**Comment by jenshannoschwalm** [comment]
@s7habo i have renamed the title of this issue to keep attention on this, ok?

@s7habo @SoupyGit i think i don't get "what exactly is the brightness issue" :-)

The module works in UC22-HSB space so the brightness - and also it's correction - is not linear, the visualizing mask shows changes in linear space though. Do you/we want more range for brightness correction? Or elaborate please ...

**Comment by s7habo** [comment]
> Or elaborate please ...

Brightness function seems to follow a kind of "s" curve. This means that in an area with the same color but different brightness, the **contrast is increased**. 


To illustrate, the brightness function follows...

![grafik](https://github.com/darktable-org/darktable/assets/38165484/17ff9236-7435-47b4-a863-13294cf7a706)


... a curve like this:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/6585b8af-ba0e-4509-91bc-3048a59e0922)

If this were a linear function, you would have something like this:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/7e573b91-4540-4531-8803-7dafdf1ff192)


And that's what you would have expected as a user.

And because that's not the case...

>  Do you/we want more range for brightness correction? 

 ...the answer is definitely yes, but how much depends on how/if this relates to increasing the contrast.

**Comment by jenshannoschwalm** [comment]
Aah, got it. Indeed the brightness correction follows the s-shaped weighing curve in a slightly different way atm, you need more saturation to have any change in brightness. Knowing that and working on it.

**Comment by jenshannoschwalm** [comment]
Yes that could be done. I take your word about usability VS user interface here. So make up your mind, yes or no :-)

**Comment by s7habo** [comment]
> So make up your mind, yes or no :-)

I say yes and am immediately on the hunt for a good example. :point_up:

**Comment by s7habo** [comment]
Here is one:

Before:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/ec41ec2a-7649-464e-83e0-827fb0b5ee13)


After:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/2f4e0548-cf20-4e05-b819-930af36f9e67)

And now I know why I thought of that. We had the discussion here about **Fuji color chrome effect**:

https://discuss.pixls.us/t/what-is-color-chrome-effect-and-how-to-achieve-it/18760/6?u=s7habo

Roughly speaking, this effect darkens the intense color areas to create an additional contrast that is lost with strong saturation.

**Comment by s7habo** [comment]
Here are two more examples:

Note how the texture of the leaves is enhanced here:

Before:


![grafik](https://github.com/darktable-org/darktable/assets/38165484/73ea6f31-00e0-4168-96f4-b004ebe776c6)


After:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/8d288784-0eea-43b9-bac0-c81c2e86a978)

Here the car gets nice contrasts:

Before: 


![grafik](https://github.com/darktable-org/darktable/assets/38165484/603dde99-ae12-4551-81b5-f9da03e79fc4)


After:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/7a7d4dad-844f-4d1f-ae98-c90f2abcc4d7)

**Comment by s7habo** [comment]
I have one more idea that I would like to get rid of and then I'll stop because it belongs in the feature request section:

If you can have the same weight slider by saturation also for the saturation function itself (stronger saturation for already saturated colors vs. even saturation) you would also have what "vibrance"  function does in color balance rgb module.

This would give both saturation and brightness functions an additional - let's call it - "vibrance" slider. I can imagine a lot of possible uses for this.  But we can talk about that later.

**Comment by s7habo** [comment]
Excellent work @jenshannoschwalm ! With your latest contribution, the color equalizer is slowly becoming the Swiss army knife for dealing with colors. :)

Here's an example right away:

In this photo...

![grafik](https://github.com/darktable-org/darktable/assets/38165484/26161298-e109-4369-985c-d52f528c999a)

...we wanted to darken the skin. With a very low effect radius, however, the skin looks very rough:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/76ce1b15-2a7d-4d34-adf7-6787a9f9c1fe)


To counter this, we now increase the **effect radius**. The skin now looks nice, but the fur on the hat has lost its local contrast:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/ba00051e-7edf-4578-b064-8f9b3516c49f)

If we now also increase **saturation threshold** to increase **saturation restriction**, the fur of the cap regains its local contrast because it is excluded from the darkening. Face has more contrast and the neck is cleanly separated from the environment, which also **prevents haloing of guided filter**. This is really a very nice additional control:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/00083d2e-8e4e-43e9-822f-185e70343a4f)

The new saturation threshold function is a **huge** improvement in my opinion!

I have two questions in this regard. 

1. The slider is limited to -/+ 10% Can the radius of action be increased to -/+30% or more? Or one leave it as it is, but offer the possibility to enter higher (or lower) values via keyboard if necessary.
2. The transitions between saturated and non-saturated areas are a bit harsh. Could this also be controlled somehow with the guided filter?

**Comment by jenshannoschwalm** [comment]
> I have two questions in this regard.
> 
>     1. The slider is limited to -/+ 10% Can the radius of action be increased to -/+30% or more? Or one leave it as it is, but offer the possibility to enter higher (or lower) values via keyboard if necessary.
> 
>     2. The transitions between saturated and non-saturated areas are a bit harsh. Could this also be controlled somehow with the guided filter?

Ad 1. Yes, although the percent value is not technically correct. With current code we can set the saturation threshold to values between 0 and ~25% (at 0 we have the default) I just thought this range would be sufficient :-) Let me know after further testing if this should be expanded, i wouldn't expect any problems from changing this.

Ad 2. Principally yes. The current weighing function is pretty steep around the threshold giving a good on/off effect. We could do that smoother of course but we reduce control of "colors should not bleed into achromatic areas" which was pretty bad in the original code. That would require an additional slider again ... something i would not like too much.

So, yes - both things could be done without big problems. As you compile yourself you might try this:

Ad 1: See line 102, change this to
```
  float threshold;          // $MIN: -0.1 $MAX: 0.2 $DEFAULT: 0.0 $DESCRIPTION: "saturation threshold"
```

Ad 2: Line 356, change the -60.0 to -30.0 for a smoother transition.
```
    satweights[i+SATSIZE] = (float)(1.0 / (1.0 + exp(-60.0 * val)));
```

**Comment by s7habo** [comment]
> So, yes - both things could be done without big problems. As you compile yourself you might try this:

Unfortunately, I'm not that good. I only compile master version according to idiot-proof instructions for Ubuntu which I just blindly follow. :)


As for the first point, I'm already pretty sure that the range can be extended. 

As for the second point, if it requires a new slider, I will also say that we don't need to overdo it. It makes sense to wait and see.

**Comment by s7habo** [comment]
And now, after the  https://github.com/darktable-org/darktable/pull/16415 :

"Barbie doll" (saturation threshold -10%) vs. natural saturation gradient in the face (saturation threshold + 18%):

![grafik](https://github.com/darktable-org/darktable/assets/38165484/d806d683-870f-439a-9794-979d113c09c8)

Uniform (saturation threshold -10%) vs. "Kodachrome 64" saturation distribution (saturation threshold + 18%):

![grafik](https://github.com/darktable-org/darktable/assets/38165484/bf5314d3-43c9-43b3-9b5e-4738c7683da7)

I can well imagine that analog film fanatics will now be very happy and a lot of presets will be circulating in the forums when the new darktable version comes out. :)

And for me it's now time for extensive testing.

Thank you  @jenshannoschwalm !

**Comment by jenshannoschwalm** [comment]
> And for me it's now time for extensive testing.

Waiting for feedback :-)

**Comment by s7habo** [comment]
> Speaking of presets, if during your testing you have some nice (and generic) one to share do not hesitate, I'll be happy to integrate them.

[colorequal_Kodachrome 64 like .dtpreset.zip](https://github.com/darktable-org/darktable/files/14535924/colorequal_Kodachrome.64.like.dtpreset.zip)

The preset is made according to the "Kodachrome 64" LUT from the RawTherapee collection here:

http://rawtherapee.com/shared/HaldCLUT.zip

In the collection is also the generic version, but this version requires additionally the shift of the white point in highlights towards magenta, which is not possible with CE module.

The preset works best in combination with the "smooth" preset in sigmoid, but can also be used with filmic.

The overall saturation can be adjusted with the color balance rgb module if required.  However, I recommend reducing the saturation in highlights as this allows you to achieve the natural saturation distribution that also occurs in LUT.

Here are a few examples (original on the left, "Kodachrome 64 like" preset on the right):

![grafik](https://github.com/darktable-org/darktable/assets/38165484/8053b498-c28d-4c38-a375-0980f30d51c7)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/fa7eaeed-e484-4a69-af0e-3edb57ec9a51)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/4c105cbb-ccdc-42ae-89ba-1d94857bcc3e)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/69cf4f60-591d-4235-b1cb-f44229f21fbc)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/e6ae0b76-9b97-4849-a6d1-45f3a82a7196)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/deceebea-227d-4980-8d25-16a9586b4d13)

I will offer some other presets soon.

**Comment by s7habo** [comment]
[colorequal_Kodak portra 400 like.dtpreset.zip](https://github.com/darktable-org/darktable/files/14549905/colorequal_Kodak.portra.400.like.dtpreset.zip)

This preset is based on "Kodak Portra UC1" from HaldCLUT collection and provides a nice color balance for most photos.

The subtle separation of cold and warm tones (shifting green and blue towards cyan and red and yellow towards orange) gives a good color contrast on the one hand and a nice color gradient for skin tones on the other.

This preset expects the saturation in general most in mid tones. Accordingly, the saturation in shadows and highlights should be reduced slightly with color balance rgb.

Here are a few examples (original on the left, "Kodak portra like" preset on the right):

![grafik](https://github.com/darktable-org/darktable/assets/38165484/c460f298-e879-4aa7-8f65-156ff2cbc9ce)


![grafik](https://github.com/darktable-org/darktable/assets/38165484/9cb186af-6778-4539-906b-e1c64b599db8)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/bcc1c605-b188-4840-a8d0-e91b53cac9c3)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/b980821b-ded3-4c8c-b1bf-0a8b2f2370bc)

**Comment by s7habo** [comment]
[Orange & Teal.dtpreset.zip](https://github.com/darktable-org/darktable/files/14550725/Orange.Teal.dtpreset.zip)

Very exploited but still very popular color style. You can use it like this or adjust its intensity with the opacity slider. 

Here are a few examples (original on the left, Orange & Teal" preset on the right):

![grafik](https://github.com/darktable-org/darktable/assets/38165484/fb005b9d-0855-46ff-96d3-9a936bc7ea18)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/0403b39b-58b1-4cb0-b338-d2d960515b9a)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/0a8638fd-aa29-4f39-ad9f-d6aa8e0adffd)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/9145a7a3-43eb-4c68-987b-6d52fdbd93aa)

**Comment by s7habo** [comment]
[bleach bypass.dtpreset.zip](https://github.com/darktable-org/darktable/files/14551089/bleach.bypass.dtpreset.zip)

A visual effect used primarily in the film industry in which the bleaching process is partially or completely skipped during color film processing. 

A famous example is the film "Saving Private Ryan" by Spielberg.

The result is a de-saturated and high-contrast photo that is on the verge of black and white, which shifts the focus to the form without completely sacrificing the color:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/e6a970b0-dff4-4e56-b0f5-5a1eea4d5975)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/e820e4d5-e762-4074-9288-6667caced83f)

![grafik](https://github.com/darktable-org/darktable/assets/38165484/1d556292-5f7f-46d6-8b0d-96f911b248d4)

Due to the fact that the colors are not very dominant, a very nice color mood can be achieved by additionally changing the global color tone with global offset in color balance rgb module:

![grafik](https://github.com/darktable-org/darktable/assets/38165484/abb26538-5de8-4f80-a6d9-dd928cbbdf0a)

**Comment by s7habo** [comment]
@jenshannoschwalm  what i noticed is that saturation threshold goes from -10% to +20%. why not from **-20%** to +20%. If this also affects the de-saturated areas that could also be an advantage.

So, if there are no technical concerns, it would be interesting to know how the module behaves when the saturation goes below -10%. :)

**Comment by jenshannoschwalm** [comment]
The range of -10% to 20% is fully sufficient as the saturation threshold is **relative to default** resulting in saturation inflection points between -0.05 to 0.6. (You can see the curve via the masking buttons as you know)

@s7habo @TurboGit: We might concider though: 
Currently we use a non-linear scale for threshold and could redefine the threshold slider
1. to be in linear 0 to 60% range, this would result in maths changing so breaking current edits
2. to be in non-linear 0 to 30% range, this could be done more easily without breaking edits

**Comment by jenshannoschwalm** [comment]
With today's master, is there still need for the guiding filter toggle on/off?

**Comment by s7habo** [comment]
> With today's master, is there still need for the guiding filter toggle on/off?

I will suggest a few days of testing and then we can decide.
I would like to try it out on some photos first.


In any case, the latest version is great! Good work @jenshannoschwalm !
And also many thanks for the quick integration into the master @TurboGit !

**Comment by jenshannoschwalm** [comment]
@TurboGit @s7habo again, do we still need the guiding toggle button?

**Comment by s7habo** [comment]
I use color equalizer often and always turn it on and off to see the differences. With some photos there are indeed differences. These are usually the differences in the sharpness of the image when the guided filter is switched on.
I would also like to play with it a little longer.

**Comment by jenshannoschwalm** [comment]
Okidoki
